clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -307,6 +308,7 @@ public static byte requiredMagicForVersion(short produceRequestVersion) {
                 return RecordBatch.MAGIC_VALUE_V1;
 
             case 3:
+            case 4:
```

Comment:
No. This is the only improvement in my mind that I haven't made in this patch. I mentioned this in a comment after my last commit yesterday.

As of this patch, producer will still send ProduceRequest with version 3 if the message magic value is 2. This is because the newly-added produce version 4 is incompatible with 0.11 broker. As of now our producer determines the request version based on the minimum magic value in the data instead of using the ApiVersionsRequest to negotiate the this with the broker. This wasn't a problem in the latest Kafka but this is causing problem for patches like KIP-112 which wants to bump up ProducRequest version without bumping up the magic value. I don't think this is a critical problem for KIP-112 since NotLeaderForPartitionException in the response is a reasonable workaround. Can I fix this issue in a followup patch later?